File: pari3f.py
def detect_loop(self):
    slow_ptr = self.head
    fast_ptr = self.head
    
    while fast_ptr and fast_ptr.next:
        slow_ptr = slow_ptr.next
        fast_ptr = fast_ptr.next.next
        if slow_ptr == fast_ptr:
            return True
            
    return False

llist = SinglyLinkedList()
llist.insert_at_end(10)
llist.insert_at_end(20)
llist.insert_at_end(30)
llist.insert_at_end(40)
llist.insert_at_end(50)

llist.head.next.next.next.next.next = llist.head.next.next

print("Detecting loop in a looped list:", llist.detect_loop())

llist2 = SinglyLinkedList()
llist2.insert_at_end(10)
llist2.insert_at_end(20)
llist2.insert_at_end(30)
print("Detecting loop in a non-looped list:", llist2.detect_loop())

File: pari3f.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
struct Node {
    int data;
    struct Node* next;
};
bool detectLoop(struct Node* head) {
    struct Node *slow_p = head, *fast_p = head;

    while (slow_p && fast_p && fast_p->next) {
        slow_p = slow_p->next;
        fast_p = fast_p->next->next;
        if (slow_p == fast_p) {
            return true;
        }
    }
    return false;
}
int main() {
    struct Node* head = (struct Node*)malloc(sizeof(struct Node));
    head->data = 1;
    head->next = (struct Node*)malloc(sizeof(struct Node));
    head->next->data = 2;
    head->next->next = (struct Node*)malloc(sizeof(struct Node));
    head->next->next->data = 3;
    head->next->next->next = (struct Node*)malloc(sizeof(struct Node));
    head->next->next->next->data = 4;
    head->next->next->next->next = head->next;
    printf("Does the list have a loop? %s\n", detectLoop(head) ? "Yes" : "No");
    struct Node* head2 = (struct Node*)malloc(sizeof(struct Node));
    head2->data = 1;
    head2->next = NULL;
    printf("Does the second list have a loop? %s\n", detectLoop(head2) ? "Yes" : "No");
    
    return 0;
}
